crates/bevy_tasks/src/single_threaded_task_pool.rs

```diff
@@ -295,21 +305,32 @@ impl<'scope, 'env, T: Send + 'env> Scope<'scope, 'env, T> {
     ///
     /// For more information, see [`TaskPool::scope`].
     pub fn spawn_on_scope<Fut: Future<Output = T> + 'scope + MaybeSend>(&self, f: Fut) {
-        let result = ScopeResult::<T>::default();
-        self.results.borrow_mut().push(result.clone());
+        // increment the number of pending tasks
+        let pending_tasks = self.pending_tasks;
+        pending_tasks.replace_with(|i| *i + 1);
+
+        // add a spot to keep the result, and record the index
+        let results_ref = self.results_ref;
+        let mut results = results_ref.borrow_mut();
+        let task_number = results.len();
+        results.push(None);
+        drop(results);
+
+        // create the job closure
         let f = async move {
-            let temp_result = f.await;
-
-            crate::cfg::std! {
-                if {
-                    result.borrow_mut().replace(temp_result);
-                } else {
-                    let mut lock = result.lock().unwrap_or_else(PoisonError::into_inner);
-                    *lock = Some(temp_result);
-                }
-            }
+            let result = f.await;
```

Comment:
I'm not super familiar with async; but if `f` panics, then `pending_tasks` wouldn't get decremented and the scope would never return, no?

Or is that not a problem because this is single-threaded so we don't catch_unwind panics?

Reply:
I've changed it so that the scope uses it's own executor. This means panics will always propagate, as far as I can tell.